src/mlpack/methods/kde/kde.hpp

```diff
@@ -0,0 +1,245 @@
+
+  /**
+   * Initialize KDE object with the default Kernel and Metric parameters.
+   * Relative error tolernce is initialized to 1e-6, absolute error tolerance
+   * is 0.0 and uses a depth-first approach.
+   */
+  KDE();
+
+  /**
+   * Initialize KDE object using the default Metric parameters and a given
+   * Kernel bandwidth (<b>only for kernels that require a bandwidth and are
+   * constructed like kernel(bandwidth)</b>).
+   *
+   * @param bandwidth Bandwidth of the kernel.
+   * @param relError Relative error tolerance of the model.
+   * @param absError Absolute error tolerance of the model.
+   * @param breadthFirst Whether the tree should be traversed using a
+   *        breadth-first approach.
+   */
+  KDE(const double bandwidth,
+      const double relError = 1e-6,
+      const double absError = 0,
+      const bool breadthFirst = false);
+
+  /**
+   * Initialize KDE object using custom instantiated Metric and Kernel objects.
+   *
+   * @param metric Instantiated metric object.
+   * @param kernel Instantiated kernel object.
+   * @param relError Relative error tolerance of the model.
+   * @param absError Absolute error tolerance of the model.
+   * @param breadthFirst Whether the tree should be traversed using a
+   *        breadth-first approach.
+   */
+  KDE(MetricType& metric,
+      KernelType& kernel,
+      const double relError = 1e-6,
+      const double absError = 0,
+      const bool breadthFirst = false);
+
+  /**
+   * Construct KDE object as a copy of the given model. This may be
+   * computationally intensive!
+   *
+   * @param other KDE object to copy.
+   */
+  KDE(const KDE& other);
+
+  /**
+   * Construct KDE object taking ownership of the given model.
+   *
+   * @param other KDE object to take ownership of.
+   */
+  KDE(KDE&& other);
+
+  /**
+   * Copy a KDE model.
+   *
+   * Use std::move if the object to copy is no longer needed.
+   *
+   * @param other KDE model to copy.
+   */
+  KDE& operator=(KDE other);
+
+  /**
+   * Destroy the KDE object. If this object created any trees, they will be
+   * deleted. If you created the trees then you have to delete them yourself.
+   */
+  ~KDE();
+
+  /**
+   * Trains the KDE model. It builds a tree using a reference set.
+   *
+   * Use std::move if the reference set is no longer needed.
+   *
+   * @param referenceSet Set of reference data.
+   */
+  void Train(MatType referenceSet);
+
+  /**
+   * Trains the KDE model. Sets the reference tree to an already created tree.
+   *
+   * @param referenceTree New already created reference tree.
+   */
+  void Train(Tree* referenceTree);
+
+  /**
+   * Estimate density of each point in the query set given the data of the
+   * reference set. The result is stored in an estimations vector.
+   *
+   * - Dimension of each point in the query set must match the dimension of each
+   *   point in the reference set.
+   *
+   * - Use std::move if the query set is no longer needed.
```

Comment:
Do we need to add another overload so that works right?

Reply:
Solved in 15e0127